akka-discovery-azure-api/src/main/scala/akka/discovery/azureapi/AzureRbacAKSServiceDiscovery.scala

```diff
@@ -0,0 +1,270 @@
+            ports <- container.ports.toSeq
+            port <- ports
+            if port.name.contains(name)
+          } yield Some(port.containerPort)
+      }
+    } yield {
+      val hostOrIp = if (rawIp) ip else s"${ip.replace('.', '-')}.$podNamespace.pod.$podDomain"
+      ResolvedTarget(
+        host = hostOrIp,
+        port = maybePort,
+        address = Some(InetAddress.getByName(ip))
+      )
+    }
+
+  final class KubernetesApiException(msg: String) extends RuntimeException(msg) with NoStackTrace
+
+  final case class KubernetesSetup(namespace: String, ctx: HttpsConnectionContext)
+}
+
+class AzureRbacAKSServiceDiscovery(implicit system: ActorSystem) extends ServiceDiscovery {
+  private val http = Http()
+
+  private val settings = Settings(system)
+
+  private val log = Logging(system, classOf[AzureRbacAKSServiceDiscovery])
+
+  log.debug("Settings {}", settings)
+
+  private def fetchToken: Future[AccessToken] =
+    AzureRbacAKSServiceDiscovery.azureDefaultCredential
+      .getToken(AzureRbacAKSServiceDiscovery.accessTokenRequestContext.addScopes(settings.entraServerId))
+      .toFuture
+      .asScala
+
+  private val kubernetesSetup = {
+    implicit val blockingDispatcher: ExecutionContext = system.dispatchers.lookup(DefaultBlockingDispatcherId)
+    for {
+      namespace: String <- Future {
+        settings.podNamespace
+          .orElse(readConfigVarFromFilesystem(settings.podNamespacePath, "pod-namespace"))
+          .getOrElse("default")
+      }
+      httpsContext <- Future(clientHttpsConnectionContext())
+    } yield {
+      KubernetesSetup(namespace, httpsContext)
+    }
+  }
+
+  import system.dispatcher
+
+  private def accessToken(retries: Int = 3): Future[AccessToken] = {
+    fetchToken.flatMap { token =>
+      if (token.isExpired && retries > 0) {
+        accessToken(retries - 1)
+      } else if (token.isExpired) {
+        Future.failed(new RuntimeException("Failed to fetch a valid token after multiple attempts"))
+      } else {
+        Future.successful(token)
+      }
+    }
+  }
+
+  override def lookup(lookup: Lookup, resolveTimeout: FiniteDuration): Future[ServiceDiscovery.Resolved] = {
+    val selector = settings.podLabelSelector(lookup.serviceName)
+
+    for {
+      ks <- kubernetesSetup
+
+      at <- accessToken()
+
+      request <- {
+        log.info(
+          "Querying for pods with label selector: [{}]. Namespace: [{}]. Port: [{}]",
+          selector,
+          ks.namespace,
+          lookup.portName
+        )
+
+        optionToFuture(
+          // FIXME | remove .get
+          podRequest(at.getToken, ks.namespace, selector),
+          s"Unable to form request; check Kubernetes environment (expecting env vars ${settings.apiServiceHostEnvName}, ${settings.apiServicePortEnvName})"
+        )
+      }
+
+      response <- http.singleRequest(request, ks.ctx)
+
+      entity <- response.entity.toStrict(resolveTimeout)
+
+      podList <- {
```

Comment:
Consider splitting this out in its one private function to make the function easier to read.

Reply:
I agree, any `<- { ... multiline block ... }` is a good sign it should have been a function or method

Reply:
Note that he inherited some style from the existing `KubernetesApiServiceDiscovery`, which was probably used as starting point. Not a reason for keeping that style, but just saying.